Sync stack description (#132)
ubclaunchpad/docs: resources/groundwork/web.md

### Sync Stack

| Pros | Cons |
|------|------|

> | Both Frontend and Backend are in one language (TypeScript). Express (Node.js) is low level, meaning it gives you a lot of control, allows for customizability, and is generally easy to debug. | React is not the easiest frontend framework to learn |

Reviewer: I mentioned this in #131, but I think we can omit specific technologies in this overview - it should be covered in more detail further down :) Maybe something high-level like "shared language means easier onboarding for inexperienced members"

> Redis is a simple, light weight key value storage database. Accountless means much less overhead and initial set up. | Complex data structures and data relationships may not fit well with a simple key value store. Accountless means it can be difficult to “attach” information to a specific user. |

Reviewer: ```suggestion
Going "accountless" means much less overhead and initial set up, and skips the need to handle user accounts and credentials. | Complex data structures and data relationships may not fit well with a simple key value store. "Accountless" means it can be difficult to “attach” information to a specific user. |
```

**Frontend**: React.js (TypeScript)

- Frameworks and libraries
  - Material UI: [Material UI](https://material-ui.com/) is React UI framework for easily building React components.
- Tooling
  - [create-react-app](https://reactjs.org/docs/create-a-new-react-app.html) is a simple way to setup the frontend.
- Considerations
  - Pro: React is currently (2020) the most popular frontend framework. This means that it has a lot of community support (eg. stack overflow), and large number of libraries/packages which can help with your project.

>   - Pro + Con: React is unopinionated - meaning the build and structure of your applications is up to you. For example, you can write your components in either class components or functional components (taking advantage of hooks), each has it’s own set up advantages, it’s up to you to decide which is best for your application

Reviewer: ```suggestion
  - Pro + Con: React is unopinionated - meaning the build and structure of your application is up to you. For example, you can write your components in either class components or functional components. Each approach has its own advantages and disadvantages, so it's up to you to decide which is best for your project.
```

I understand why you might bring up hooks, but there's also not enough context here to understand what that means - I'm a bit torn on the class components vs functional components comparison too, since to a complete beginner it doesn't make too much sense. I wonder if it might be better to say something like:

```suggestion
  - Pro + Con: React is unopinionated - meaning the build and structure of your application is up to you. There are many ways to do everything in React, and each approach has its own advantages and disadvantages, so it's up to you to decide which is best for your project.
```

- References
  - [TypeScript resources](/resources/languages.md#typescript)

**Backend**: Node.js (TypeScript)
- Frameworks and libraries

>   - Express: Express.js is a popular Node.js server-side framework. It provides a set of features that makes backend developing easier. Like React, it is unopinioated and there are many ways to solve the same problem. [Express app generator](https://expressjs.com/en/starter/generator.html) can quickly set up application boilerplate

Reviewer: ```suggestion
  - [Express](https://expressjs.com/) is a popular Node.js server-side framework. It provides a set of features that makes backend developing easier. Like React, it is unopinioated and there are many ways to solve the same problem. [Express app generator](https://expressjs.com/en/starter/generator.html) can quickly set up application boilerplate
```

>   - Logging: [`Pino`](https://github.com/pinojs/pino) is a low overhead multi-purpose Node.js logger

Reviewer: ```suggestion
  - Logging: [`Pino`](https://github.com/pinojs/pino) is a low-overhead structured Node.js logger.
```

I think the key thing about Pino is structured logging, which will be new to most people who have only ever logged out strings to debug things in their application - temped to add more details about structured logging here, but I think it deserves an overview of its own 😅

**Datastore**: Redis

- Frameworks and libraries

>   - [`Redis`](https://redis.io/) is an open source NoSQL in-memory data store which can be used as a database and/or a cache

Reviewer: ```suggestion
  - [`Redis`](https://redis.io/) is a versatile "key-value" store (like a big hashmap) which can be used as a database and/or a cache
```

(friendlier wording)

- Considerations

>   - Pro: The [community](https://redis.io/) version of Redis is free

Reviewer: ```suggestion
```

Imo it should be a given that all the tools we choose here are free :) no proprietary tech plz

However, one thing that might be useful to point out is that there are a few options for free redis _deployments_:

* heroku (25mb) https://elements.heroku.com/addons/heroku-redis
* redis cloud (30mb) https://redislabs.com/redis-enterprise-cloud/pricing/

Reviewer: tracking in https://github.com/ubclaunchpad/docs/issues/133
